tests/phpunit/CRM/Contact/BAO/GroupContactCacheTest.php

```diff
@@ -188,6 +159,100 @@ public function assertCacheMatches($expectedContactIds, $groupId) {
     $this->assertEquals($expectedContactIds, $actualContactIds);
   }
 
+  /**
+   * Test the opportunistic refresh cache function does not touch non-expired entries.
+   */
+  public function testOpportunisticRefreshCacheNoChangeIfNotExpired() {
+    list($group, $living, $deceased) = $this->setupSmartGroup();
+    $this->callAPISuccess('Contact', 'create', array('id' => $deceased[0]->id, 'is_deceased' => 0));
+    $this->assertCacheMatches(
+      array($deceased[0]->id, $deceased[1]->id, $deceased[2]->id),
+      $group->id
+    );
+    CRM_Contact_BAO_GroupContactCache::opportunisticCacheFlush();
+
+    $this->assertCacheNotRefreshed($deceased, $group);
+  }
+
+  /**
+   * Test the opportunistic refresh cache function does refresh stale entries.
+   */
+  public function testOpportunisticRefreshChangeIfCacheDateFieldStale() {
+    list($group, $living, $deceased) = $this->setupSmartGroup();
+    $this->callAPISuccess('Contact', 'create', array('id' => $deceased[0]->id, 'is_deceased' => 0));
+    CRM_Core_DAO::executeQuery('UPDATE civicrm_group SET cache_date = DATE_SUB(NOW(), INTERVAL 7 MINUTE) WHERE id = ' . $group->id);
```

Comment:
So, after repeated facepalms I thought about why I had made such a dumb error on the time ... and I realised I'd used copy & paste. So, I grepped & found a case sensitive match on 'Ymdhis' found a handful of matches - 6 of which may indicate a problem